File: assignment1/src/Main.java
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        Graph graph = new Graph("graph.fmi");

        long startTime = System.currentTimeMillis();
        int numComponents = graph.countWeaklyConnectedComponents();
        long endTime = System.currentTimeMillis();

        System.out.println("Weakly connected components: " + numComponents);
        System.out.println("Time: " + (endTime - startTime) + " ms");

        Dijkstra dijkstra = new Dijkstra(graph);

        BufferedReader br = new BufferedReader(new FileReader("queries.txt"));
        StringBuilder result = new StringBuilder();
        String[] queryLine;
        double distance;
        for (String query = br.readLine(); query != null; query = br.readLine()) {
            if (query.trim().isEmpty() || query.startsWith("#")) continue;

            queryLine = query.split(" ");
            startTime = System.currentTimeMillis();
            distance = dijkstra.shortestPath(Integer.parseInt(queryLine[0]), Integer.parseInt(queryLine[1]));
            endTime = System.currentTimeMillis();
            result.append(queryLine[0]).append(" ").append(queryLine[1]).append(" ").append(distance).append(" ").append(endTime - startTime).append("ms\n");
        }

        BufferedWriter writer = new BufferedWriter(new FileWriter("result.txt", false));
        writer.write(result.toString());
        writer.close();
    }
}

File: assignment1/src/main.rs
use std::error::Error;
use std::time::Instant;
use crate::dijkstra::Dijkstra;

mod graph;
mod dijkstra;
mod file;

fn main() -> Result<(), Box<dyn Error>> {
    let graph = file::parse("graph.fmi")?;
    let start = Instant::now();
    let wcc = graph.get_weakly_connected_components();
    println!("Number of weakly connected components: {}, Time: {}ms", wcc, start.elapsed().as_millis());
    
    let mut dijkstra = Dijkstra::new(&graph);
    let queries = file::read_queries("queries.txt")?;
    let mut results: Vec<(u32, u32, Option<u32>, u128)> = Vec::with_capacity(queries.len());
    for (source, target) in queries {
        let start = Instant::now();
        let result = dijkstra.query(source, target);
        results.push((source, target, result, start.elapsed().as_millis() as u128));
    }

    file::write_result("results.txt", &results)?;

    Ok(())
}
